mirrord/kube/src/api.rs

```diff
@@ -0,0 +1,114 @@
+use actix_codec::{AsyncRead, AsyncWrite};
+use async_trait::async_trait;
+use futures::{SinkExt, StreamExt};
+use k8s_openapi::NamespaceResourceScope;
+use kube::{Api, Client};
+use mirrord_progress::Progress;
+use mirrord_protocol::{ClientCodec, ClientMessage, DaemonMessage};
+use tokio::{
+    net::{TcpStream, ToSocketAddrs},
+    sync::mpsc,
+};
+use tracing::error;
+
+use crate::error::{KubeApiError, Result};
+
+mod container;
+#[cfg(feature = "env_guard")]
+mod env_guard;
+pub mod kubernetes;
+mod runtime;
+
+pub(crate) fn get_k8s_api<K>(client: &Client, namespace: Option<&str>) -> Api<K>
+where
+    K: kube::Resource<Scope = NamespaceResourceScope>,
+    <K as kube::Resource>::DynamicType: Default,
+{
+    if let Some(namespace) = namespace {
+        Api::namespaced(client.clone(), namespace)
+    } else {
+        Api::default_namespaced(client.clone())
+    }
+}
+
+pub(crate) fn wrap_raw_connection(
+    stream: impl AsyncRead + AsyncWrite + Unpin + Send + 'static,
+) -> Result<(mpsc::Sender<ClientMessage>, mpsc::Receiver<DaemonMessage>)> {
+    let mut codec = actix_codec::Framed::new(stream, ClientCodec::new());
+
+    let (in_tx, mut in_rx) = mpsc::channel(1000);
+    let (out_tx, out_rx) = mpsc::channel(1000);
+
+    tokio::spawn(async move {
+        loop {
+            tokio::select! {
+                Some(msg) = in_rx.recv() => {
```

Comment:
I think it's worth covering the specific case of None from each side to know who disconnected first.